FurkannOzbek/application-hub
# [FEATURE] Implement Token Refresh

# [FEATURE] Implement Token Refresh 

---

## Description

This PR implements a token refresh mechanism and enhances the overall authentication system, improving security and user experience. The changes include automatic token refresh before expiration.

Impact:
- Automatic token refresh when approaching expiration (10 minutes threshold)
- Consolidated authentication utilities
- Consolidated authentication API calls
- Improved error handling and status codes


---

## Changes Made

### Backend Changes
- Added new `/refresh-token` endpoint for token renewal
- Consolidated authentication utilities into `auth.js`
- Improved token invalidation and cleanup processes
- Standardized token extraction from headers
- Updated status codes for authentication responses

### Frontend Changes
- Implemented automatic token refresh mechanism
- Added JWT decode functionality for token expiration checking
- Enhanced Authentication context with token refresh logic
- Consolidated authentication API calls

### Dependencies
Added `jwt-decode` package for frontend token handling

---

## Testing Steps

1. Token Refresh Flow:
- Login to the application
- Wait for a token to approach expiration
- Verify automatic token refresh occurs
- Confirm session continues without interruption
- 
2. Logout Flow:
- Test logout functionality
- Verify token invalidation
- Confirm redirect to login page
- Verify the existence of an invalidated token in the database
